### ISSUE TITLE
Handle exceptions better in tensor mechanics

### ISSUE THREAD

**Issue by tophmatthews** [idaholab/moose#16263]
## Reason
Exceptions are used to ensure sanity in calculations during some tensor mechanics models. If these exceptions are thrown anywhere other than `nonlinear` and `linear` (I think those are the only two?), then the run is killed rather than a timestep cut. These exceptions can be protected by ignoring the condition where exceptions occur.

## Design
Add conditionals to protect exceptions:
```
if (_current_execute_flag == EXEC_LINEAR || _current_execute_flag == EXEC_NONLINEAR)
    if (some_condition)
      mooseException("asdf")
```

## Impact
This will avoid premature model terminations due to exceptions thrown at bad times.

**Comment by dschwen** [comment]
Where do you observe the exceptions being thrown? On timestep end for output? Why are material exceptions triggered anywhere after the solve has successfully finished?

**Comment by tophmatthews** [comment]
initial too

**Comment by tophmatthews** [comment]
I'm not sure why they're called at initial or timestep end, but it happens enough that I started putting guards on all my exceptions where I don't want it to error out. I see it mostly on initial, which if it continues on in the first linear/nonlinear, then so what?

**Comment by tophmatthews** [comment]
I would love to see a better handling of mooseException, but I've been talking (whining?) about it long enough that it looks like there is no easy solution

**Comment by dschwen** [comment]
But what should happen on initial? You cannot cut a timestep. Just ignoring the error seems odd. You'd potentially get a wrong IC.

**Comment by tophmatthews** [comment]
I mean timestep begin. Fair point about initial though.

What is a better solution? I am constantly hitting this problem, and it's preventing work.

Wrap exception in something smarter that gives you an option to kill or continue?

**Comment by dschwen** [comment]
Closed by #23011
